### ISSUE TITLE
Griffin-Lim for CQT

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#912]
#### Description

The GL implementation added in 0.7 #843 operates on linear spectrograms.  At some point in the discussion, we raised the idea of having it work on CQT as well, but dropped the idea to keep things simple.

I just prototyped a CQT-based GL offline and it sounds pretty good.  It seems like a useful kind of thing to have in the library.  The only question for me is whether we wrap it into the existing implementation, or just have a separate function that's nearly identical, but with different arguments and calls `cqt/icqt` instead of `stft/istft`.  I'm leaning toward the latter, since the method is simple enough that a little duplication of code won't be catastrophic to maintain.

What do folks think?

**Comment by bmcfee** [comment]
For the record, here's a link to my implementation [notebook](https://gist.github.com/bmcfee/94bc48f5ffcd20aa4929c0642cd40ad2).  It's functional, but the API needs to be expanded to encapsulate all of the CQT parameters.

**Comment by lostanlen** [comment]
The code runs on my machine! Good job!
But the top 2-3 octaves are missing from the CQT filterbank so they are missing from the reconstruction, of course. It's difficult to judge perceptually without them.
![Screen Shot 2019-07-02 at 2 33 57 PM](https://user-images.githubusercontent.com/3943142/60537398-71f00300-9cd6-11e9-93df-c94572f41f81.png)

**Comment by bmcfee** [comment]
> But the top 2-3 octaves are missing from the CQT filterbank so they are missing from the reconstruction, of course. It's difficult to judge perceptually without them.

Sure -- that's always the case with our cqt implementation since it doesn't generally cover up to nyquist.  I think the sanest way to evaluate this is relative to `icqt` with given phase.

**Comment by lostanlen** [comment]
This works for a very narrow value of hop lengths and Q's, no?

I'm very often hitting on either

```python
/Users/vl238/librosa/librosa/core/constantq.py:657: UserWarning: hop_length=128 exceeds minimum CQT filter length=94.550.
This will probably cause unpleasant acoustic artifacts. Consider decreasing your hop length or increasing the frequency resolution of your CQT.
  'frequency resolution of your CQT.'.format(hop_length, min(lengths)))
```

or 

```
/Users/vl238/librosa/librosa/core/constantq.py:657: UserWarning: hop_length=128 exceeds minimum CQT filter length=94.550.
This will probably cause unpleasant acoustic artifacts. Consider decreasing your hop length or increasing the frequency resolution of your CQT.
  'frequency resolution of your CQT.'.format(hop_length, min(lengths)))
```

The latter appears at every iteration of Griffin-Lim. Is there a way to make the `UserWarning` print only once?

**Comment by bmcfee** [comment]
> This works for a very narrow value of hop lengths and Q's, no?

It's subject to all the limitations of our icqt implementation, so yes.



> The latter appears at every iteration of Griffin-Lim. Is there a way to make the `UserWarning` print only once?

A user can always `filterwarnings` to `'once'`: https://docs.python.org/3/library/warnings.html#warnings.filterwarnings

I don't think we should do that from within the library though, since a warning can pop up from different sources within a session, and only showing the first one would obscure that.

**Comment by keunwoochoi** [comment]
> I'm leaning toward the latter

+1, but probably under a slightly different name than `icqt`, e.g.,`pseudo-icqt` for example. Because it's an estimation. Maybe with a keyword arg like `method='gl'` or something.

**Comment by bmcfee** [comment]
> but probably under a slightly different name than `icqt`, e.g.,`pseudo-icqt`

Absolutely, it can't be `icqt` (already taken, and used as a subroutine for GL-CQT), and `pseudo-icqt` is misleading because we already have a forward pseudo-cqt that means something different.

I think `griffinlim_cqt` is the most direct name.

**Comment by keunwoochoi** [comment]
Assuming you meant `griffinlim_icqt`, sounds good to me!

**Comment by lostanlen** [comment]
how about `cqt_to_audio`?

**Comment by bmcfee** [comment]
> how about `cqt_to_audio(S, method='griffin_lim')`?

That's consistent with the mel/mfcc interfaces, but not with the current `griffinlim` interface.  I think it would be more confusing to do it this way since we already have `icqt`.

**Comment by JulianFerry** [comment]
> > The latter appears at every iteration of Griffin-Lim. Is there a way to make the `UserWarning` print only once?
> 
> A user can always `filterwarnings` to `'once'`: https://docs.python.org/3/library/warnings.html#warnings.filterwarnings
> 
> I don't think we should do that from within the library though, since a warning can pop up from different sources within a session, and only showing the first one would obscure that.

I tried using `filterwarnings` as you suggested, but this only works with "ignore", not with "once" (or "module")

[Using the example in the docs](https://librosa.github.io/librosa/generated/librosa.core.griffinlim_cqt.html), the warning message gets printed multiple times:
```
from warnings import filterwarnings
filterwarnings("once", category=UserWarning)

hop_length = 1024
y, sr = librosa.load(librosa.util.example_audio_file(), duration=5, offset=30, sr=None)
C = np.abs(librosa.cqt(y=y, sr=sr, bins_per_octave=36, n_bins=7*36, hop_length=1024))
y_inv = librosa.griffinlim_cqt(C, sr=sr, bins_per_octave=36, hop_length=1024)
```

Sorry to necro this thread, I can make this into a separate issue if you want.

**Comment by bmcfee** [comment]
I don't think this is an issue for us, except maybe documentation.  You may need to reset your warning filters before applying a new filter, but this depends on your environment (jupyter/ipython/spyder/python interpreter may all behave differently).
